aptos-move/move-examples/data_structures/sources/smart_table.move

```diff
@@ -0,0 +1,358 @@
+                let Entry {hash:_, key:_, value} = vector::swap_remove(bucket, i);
+                table.size = table.size - 1;
+                return value
+            };
+            i = i + 1;
+        };
+        abort error::invalid_argument(ENOT_FOUND)
+    }
+
+    /// Returns the length of the table, i.e. the number of entries.
+    public fun length<K, V>(table: &SmartTable<K, V>): u64 {
+        table.size
+    }
+
+    /// Return the load factor of the hashtable.
+    public fun load_factor<K, V>(table: &SmartTable<K, V>): u64 {
+        table.size * 100 / (table.num_buckets * table.bucket_size)
+    }
+
+    /// Update `split_load_threshold`.
+    public fun update_split_load_threshold<K, V>(table: &mut SmartTable<K, V>, split_load_threshold: u8) {
+        assert!(split_load_threshold <= 100 && split_load_threshold > 0, error::invalid_argument(EINVALID_LOAD_THRESHOLD_PERCENT));
+        table.split_load_threshold = split_load_threshold;
+    }
+
+    /// Update `bucket_size`.
+    public fun update_bucket_size<K, V>(table: &mut SmartTable<K, V>, bucket_size: u64) {
+        assert!(bucket_size > 0, error::invalid_argument(EINVALID_LOAD_THRESHOLD_PERCENT));
+        table.bucket_size = bucket_size;
+    }
+
+    /// Reserve `additional_buckets` more buckets.
+    fun split<K, V>(table: &mut SmartTable<K, V>, additional_buckets: u64) {
+        while (additional_buckets > 0) {
+            additional_buckets = additional_buckets - 1;
+            split_one_bucket(table);
+        }
+    }
+
+    #[test]
+    fun smart_table_test() {
+        let table = new();
+        let i = 0;
+        while (i < 200) {
+            add(&mut table, i, i);
+            i = i + 1;
+        };
+        assert!(length(&table) == 200, 0);
+        i = 0;
+        while (i < 200) {
+            *borrow_mut(&mut table, i) = i * 2;
+            assert!(*borrow(&mut table, i) == i * 2, 0);
+            i = i + 1;
+        };
+        i = 0;
+        assert!(table.num_buckets > 5, table.num_buckets);
+        while (i < 200) {
+            assert!(contains(&table, i), 0);
+            assert!(remove(&mut table, i) == i * 2, 0);
+            i = i + 1;
+        };
+        destroy_empty(table);
+    }
+
+    #[test]
+    fun smart_table_split_test() {
+        let table: SmartTable<u64, u64> = new();
+        let i = 2;
+        let level = 1;
+        while (i <= 256) {
+            assert!(table.num_buckets == i, 0);
+            assert!(table.level == level, i);
+            split_one_bucket(&mut table);
+            i = i + 1;
+            if (i == 1 << (level + 1)) {
+                level = level + 1;
+            };
+        };
+        destroy_empty(table);
+    }
+
+    #[test]
+    fun smart_table_bucket_index_test() {
+        let table: SmartTable<u64, u64> = new_with_config<u64, u64>(8, 75, 0);
+        assert!(table.level == 3, 0);
+        let i = 0;
+        while (i < 4) {
+            split_one_bucket(&mut table);
```

Comment:
This is not a good test. It's calling a private function for splitting the buckets instead of using public functions to insert elements to trigger splitting.

Reply:
I delete this one cuz it will be tested implicitly by others.